crates/matrix-sdk/src/media.rs

```diff
@@ -140,19 +140,55 @@ impl Media {
     pub async fn get_media_file(
         &self,
         request: &MediaRequest,
+        body: Option<String>,
         content_type: &Mime,
         use_cache: bool,
     ) -> Result<MediaFileHandle> {
         let data = self.get_media_content(request, use_cache).await?;
 
-        let mut suffix = String::from("");
-        if let Some(extension) =
-            mime_guess::get_mime_extensions(content_type).and_then(|a| a.first())
-        {
-            suffix = String::from(".") + extension;
-        }
+        let inferred_extension = mime2ext::mime2ext(content_type);
+
+        let body_path = body.as_ref().map(Path::new);
+        let filename = body_path.and_then(|f| f.file_name().and_then(|f| f.to_str()));
+        let filename_with_extension = body_path.and_then(|f| {
+            if f.extension().is_some() {
+                f.file_name().and_then(|f| f.to_str())
+            } else {
+                None
+            }
+        });
+
+        let file = match (filename, filename_with_extension, inferred_extension) {
+            // If the body is a file name and has an extension use that
+            (Some(_), Some(filename_with_extension), Some(_)) => {
+                // We're potentially using the same file name multiple times
+                // which might lead to tempfile() failing. Use `make` instead
+                // to avoid that
+                TempFileBuilder::new()
+                    .prefix(filename_with_extension)
+                    .rand_bytes(0)
+                    .make(|path| File::create(path))?
```

Comment:
It's probably not an issue in reality, but I wonder if this is the right approach. This would you from ever opening 2 different files with the same name at the same time. Might be better to put the file in a random subdir to avoid this?

Reply:
Yeah I think we should generate a random tempdir and put the file there with the specified name.

Reply:
Good idea! Unfortunately the implementation proved to a bit more fun than I wanted: we now generate temporary intermediary directories for files who's names might clash and we store that handle together with the `MediaFileHandle` so that it doesn't get dropped. This is transparent to the final client which just retains the media handle as a whole